openwisp_controller/config/who_is/service.py

```diff
@@ -0,0 +1,275 @@
+                            "organization_id"
+                        )[:1]
+                    )
+                )
+            except OrganizationConfigSettings.DoesNotExist:
+                # If organization settings do not exist, fall back to global setting
+                return app_settings.WHO_IS_ENABLED
+            cache.set(
+                self.ORG_SETTINGS_CACHE_KEY.format(org_pk=org_pk),
+                org_settings,
+                timeout=Config._CHECKSUM_CACHE_TIMEOUT,
+            )
+        return getattr(org_settings, "who_is_enabled", app_settings.WHO_IS_ENABLED)
+
+    def _get_who_is_info_from_db(self, ip_address):
+        """
+        For getting existing WhoIsInfo for given IP from db if present.
+        """
+        return WhoIsInfo.objects.filter(ip_address=ip_address).first()
+
+    def _need_who_is_lookup(self, initial_ip, new_ip):
+        """
+        This is used to determine if the WhoIs lookup should be triggered
+        when the device is saved.
+
+        The lookup is triggered if:
+            - The new IP address is not None.
+            - The WhoIs information is not already present or the initial IP is
+              different from the new IP.
+            - The new IP address is a global (public) IP address.
+            - WhoIs is enabled in the organization settings. (query from db)
+        """
+
+        # Check cheap conditions first before hitting the database
+        return (
+            self.is_valid_public_ip_address(new_ip)
+            and (initial_ip != new_ip or not self._get_who_is_info_from_db(new_ip))
+            and self.is_who_is_enabled
+        )
+
+    def get_device_who_is_info(self):
+        """
+        Used to get WhoIsInfo for a device if last_ip is valid public ip
+        and WhoIs is enabled.
+        """
+        ip_address = self.device.last_ip
+        if not (self.is_valid_public_ip_address(ip_address) and self.is_who_is_enabled):
+            return None
+
+        return self._get_who_is_info_from_db(ip_address=ip_address)
+
+    def trigger_who_is_lookup(self):
+        """
+        Trigger WhoIs lookup based on the conditions of `_need_who_is_lookup`.
+        Task is triggered on commit to ensure redundant data is not created.
+        """
+        if self._need_who_is_lookup(self.device._initial_last_ip, self.device.last_ip):
+            transaction.on_commit(
+                lambda: self.fetch_who_is_details.delay(
+                    device_pk=self.device.pk,
+                    initial_ip_address=self.device._initial_last_ip,
+                    new_ip_address=self.device.last_ip,
+                )
+            )
+
+    # device_pk is used when task fails to report for which device failure occurred
+    @shared_task(
+        bind=True,
+        base=WhoIsCeleryRetryTask,
+        **app_settings.API_TASK_RETRY_OPTIONS,
+    )
+    def fetch_who_is_details(self, device_pk, initial_ip_address, new_ip_address):
```

Comment:
Question: Is celery worker picking up this task? AFAIK, it only registers task which are present in `tasks,py`